### ISSUE TITLE
Pytorch not working properly (I don't know how to summarize it, see below)

### ISSUE THREAD

**Issue by davidmoseler** [pytorch/pytorch#51320]
When I have a pytorch model, I sometimes would like to extract the features before the final softmax layers or such. Here, I have a model trained and loaded from a pickle:

```
def build_model():
    model = resnet18(pretrained=True)
  
    n_features = model.fc.in_features
    n_hidden = 100
    model.fc = torch.nn.Sequential(
        torch.nn.Linear(n_features, n_hidden),
        torch.nn.ReLU(),
        torch.nn.Linear(n_hidden, 2)
    )
    
    model.to(device)
    return model

model = build_model()
model.load_state_dict(torch.load('./model.pickle'))
model.eval()
```

Then, I would suppose that the model can be rebuild from it's children:

```
modules = list(model.children())
encoder = nn.Sequential(*modules)
```

However, given a test tensor:

```
>>> x_test.shape
torch.Size([100, 3, 128, 128])
```

model(x_test) produces an output normaly, but encoder(x_test) gives RuntimeError: mat1 dim 1 must match mat2 dim 0. I don't have any idea on how to investigate it further. The error messages are quite poor.

**Comment by ezyang** [comment]
You should be getting a backtrace in this case. What is your backtrace?

**Comment by SsnL** [comment]
> Then, I would suppose that the model can be rebuild from it's children:
> modules = list(model.children())
> encoder = nn.Sequential(*modules)

This is false in general. You can guarantee that the `forward` sequentially calls the children in the same order as `children()` returns.

**Comment by davidmoseler** [comment]
@SsnL It actually does call them in the same order. I investigated layer by layer and discovered though that the last layer received an output of the format (N) and the next-to-last layer was outputting something of the format (N, 1, 1) in the restnet model. I have no idea how does that work inside of resnet. I did something like:

```
modules = list(model.children())[:-1]
encoder = nn.Sequential(*modules)
model.children()[-1](encoder(x_test).reshape(N))
```

and it worked. I don't know if that should be considered a bug in resnet. I find that weird.

**Comment by jbschlosser** [comment]
Hey @davidmoseler!

As @SsnL mentioned, in general, there is no guarantee that a module's `forward()` method calls its children in the same order as returned by `children()`. Modules have full flexibility in how computation is done in their respective `forward()` methods.

That said, as you pointed out, for the particular case of [ResNet](https://github.com/pytorch/vision/blob/09ce6758ad6ab3177f3fdec3b1bb7c7033b16712/torchvision/models/resnet.py#L230), it does happen to call them in order. For this specific case, you can swap the final linear layer with an identity layer (known as `fc` for this module) to extract features:
```python
model.fc = torch.nn.Identity()
x_test = torch.randn(100, 3, 128, 128)
output = model(x_test)
print(output.shape)   # Prints torch.Size([100, 512])
```

You may find [this tutorial](https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html) useful; it describes the fine-tuning process for torchvision models.

**Comment by SsnL** [comment]
In this case, there is reshape that does not show up in the form of a module (but as a function call in `forward`) so your new module misses it. Either way, it is not guaranteed to work.

**Comment by davidmoseler** [comment]
@SsnL Makes sense. However, how could I detect the presence of the reshape in the resnet model?
